### ISSUE TITLE
[Beta Testing] FSRS4Anki 4.0.0 Scheduler, Optimizer and Helper.

### ISSUE THREAD

**Issue by L-M-Sherlock** [open-spaced-repetition/fsrs4anki#348]
- Scheduler v4.2.1: https://github.com/open-spaced-repetition/fsrs4anki/blob/v4.2.1/fsrs4anki_scheduler.js
- Optimizer v4.2.1: https://github.com/open-spaced-repetition/fsrs4anki/blob/v4.2.1/fsrs4anki_optimizer.ipynb
- Helper add-on (has been uploaded): [fsrs4anki-helper.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/12052144/fsrs4anki-helper.zip)

**Comment by Expertium** [comment]
I suggest capping requested R between 0.75 and 0.97 in the scheduler code. This is needed to prevent the user from choosing insanely high or low values.

**Comment by L-M-Sherlock** [comment]
> This is needed to prevent the user from choosing insanely high or low values.

I agree. But the optimizer sometime suggests R=0.7. For some users with bad memory, they would not achieve 0.75.

Besides, the scheduler would still store the input R. And the helper add-on should also cap it when getting the parameters from the custom scheduling code.

**Comment by Expertium** [comment]
Sherlock,  I recommend you to run the new optimizer on all collections submitted to you via the Google form and average (weighted by the number of reviews) the optimal parameters. This would achieve 2 things:
1) A better set of initial parameters could improve convergence during optimization
2) If the user doesn't have enough data, they will just get average parameters. It should be good enough at first. I suggest introducing a hard limit - 1000 reviews, excluding same-day reviews. If user has more reviews than that, run the optimizer. Otherwise print a warning, like "Insufficient data", and print average parameters.

**Comment by Expertium** [comment]
Also, it would be great if you implemented what I suggested [here](https://github.com/open-spaced-repetition/fsrs4anki/issues/314#issuecomment-1625487475).

**Comment by Expertium** [comment]
As @user1823 suggested, the helper-add on should be compatible with both v3 and v4. In other words, it should be backward-compatible, to ensure that if someone is still using v3 their add-on won't break. Is this implemented?

**Comment by L-M-Sherlock** [comment]
> As @user1823 suggested, the helper-add on should be compatible with both v3 and v4. In other words, it should be backward-compatible, to ensure that if someone is still using v3 their add-on won't break. Is this implemented?

Of course. You can test it with v3 and v4 both.

**Comment by cjdduarte** [comment]
When I insert this new code "fsrs4anki_scheduler.js" and when running the helper this message appears below

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/42243532/753235b4-e7e2-483d-8317-ca43eccd0c8a)

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/b80de0ae-e8dd-44c4-8f27-2dfe3cf733bf)
I'm genuinely curious how I'm getting this error, considering that curve_fit should only run if there are more than 100 reviews for that grade.

**Comment by L-M-Sherlock** [comment]
> I'm genuinely curious how I'm getting this error, considering that curve_fit should only run if there are more than 100 reviews for that grade.

Could you share the related deck file with me?

**Comment by Expertium** [comment]
https://drive.google.com/file/d/1WyYRkoZZLllbESz5eCXq7hxFaHAIfrAJ/view?usp=sharing

**Comment by Expertium** [comment]
I don't know, it's a pre-made deck, I didn't make these cards. I guess whoever made them somehow messed it up, and the creation date is now the earliest date in Unix time.

**Comment by L-M-Sherlock** [comment]
> I don't know, it's a pre-made deck, I didn't make these cards. I guess whoever made them somehow messed it up, and the creation date is now the earliest date in Unix time.

OK, I will remove the filter in the next patch.

**Comment by perhydrol** [comment]
I have same question. But all cards were made by myself. I sure that all cards were created in 2023.
My deck: 
[LinearAlgebra.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/12040851/LinearAlgebra.zip)

```
ValueError                                Traceback (most recent call last)
[c:\Users\telly\Documents\project\fsrs4anki\fsrs4anki\fsrs4anki_optimizer.ipynb](file:///C:/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.ipynb) 单元格 11 in 1
      [1](vscode-notebook-cell:/c%3A/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.ipynb#X13sZmlsZQ%3D%3D?line=0) """
      [2](vscode-notebook-cell:/c%3A/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.ipynb#X13sZmlsZQ%3D%3D?line=1) w[0]: initial_stability_for_again_answer
      [3](vscode-notebook-cell:/c%3A/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.ipynb#X13sZmlsZQ%3D%3D?line=2) w[1]: initial_stability_step_per_rating
   (...)
     [16](vscode-notebook-cell:/c%3A/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.ipynb#X13sZmlsZQ%3D%3D?line=15) https://github.com/open-spaced-repetition/fsrs4anki/wiki/Free-Spaced-Repetition-Scheduler
     [17](vscode-notebook-cell:/c%3A/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.ipynb#X13sZmlsZQ%3D%3D?line=16) """
     [18](vscode-notebook-cell:/c%3A/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.ipynb#X13sZmlsZQ%3D%3D?line=17) optimizer.define_model()
---> [19](vscode-notebook-cell:/c%3A/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.ipynb#X13sZmlsZQ%3D%3D?line=18) optimizer.pretrain()
     [20](vscode-notebook-cell:/c%3A/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.ipynb#X13sZmlsZQ%3D%3D?line=19) print(optimizer.init_w)

File [c:\Users\telly\Documents\project\fsrs4anki\fsrs4anki\fsrs4anki_optimizer.py:518](file:///C:/Users/telly/Documents/project/fsrs4anki/fsrs4anki/fsrs4anki_optimizer.py:518), in Optimizer.pretrain(self)
    515 def S0_rating_curve(rating, a, b, c):
    516     return np.exp(a + b * rating) + c
--> 518 params, covs = curve_fit(S0_rating_curve, list(rating_stability.keys()), list(rating_stability.values()), sigma=1/np.sqrt(list(rating_count.values())), method='dogbox', bounds=((-15, 0.03, 0), (15, 7, 30)))
    519 print('Weighted fit parameters:', params)
    520 predict_stability = S0_rating_curve(np.array(list(rating_stability.keys())), *params)

File [c:\Users\telly\Documents\project\fsrs4anki\.conda\lib\site-packages\scipy\optimize\_minpack_py.py:820](file:///C:/Users/telly/Documents/project/fsrs4anki/.conda/lib/site-packages/scipy/optimize/_minpack_py.py:820), in curve_fit(f, xdata, ydata, p0, sigma, absolute_sigma, check_finite, bounds, method, jac, full_output, **kwargs)
    817         xdata = np.asarray(xdata, float)
    819 if ydata.size == 0:
--> 820     raise ValueError("`ydata` must not be empty!")
    822 # Determine type of sigma
    823 if sigma is not None:

ValueError: `ydata` must not be empty!
```
![图片](https://github.com/open-spaced-repetition/fsrs4anki/assets/50535400/57332681-aa57-479a-91e5-61a11344982a)


> > https://drive.google.com/file/d/1WyYRkoZZLllbESz5eCXq7hxFaHAIfrAJ/view?usp=sharing
> 
> ![image](https://user-images.githubusercontent.com/32575846/253280908-88f0fbf5-25f0-4aca-b948-6720fbef1d42.png)
> 
> I find that your cards' cids are weird. They are too small. Normally, the cid is the timestamp of the create date. In the 4.0.0, the optimizer will filter out all cards whose cid is early than 2006 (the year the first Anki version was released).我发现你的卡片的 cid 很奇怪。它们太小了。通常，cid 是创建日期的时间戳。在 4.0.0 中，优化器将过滤掉 cid 早于 2006 年（第一个 Anki 版本发布的那一年）的所有卡。

**Comment by L-M-Sherlock** [comment]
> I have same question. But all cards were made by myself. I sure that all cards were created in 2023.

Your deck only has 400+ reviews, which are not enough to optimize the model. I will add some information about that instead of error raised by the code.

**Comment by Expertium** [comment]
> Your deck only has 400+ reviews, which are not enough to optimize the model. I will add some information about that instead of error raised by the code.

Speaking of which, that's one of the reasons why I suggested [this](https://github.com/open-spaced-repetition/fsrs4anki/issues/348#issuecomment-1633937272).

**Comment by L-M-Sherlock** [comment]
> > Your deck only has 400+ reviews, which are not enough to optimize the model. I will add some information about that instead of error raised by the code.
> 
> Speaking of which, that's one of the reasons why I suggested [this](https://github.com/open-spaced-repetition/fsrs4anki/issues/348#issuecomment-1633937272).

Yeah. And I have add some useful warning in #349:

<img width="677" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/b8a93e21-3c70-4e9f-a60e-1443c00c6f69">

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/0dcf6da9-1cae-438e-b5c9-27b21e55b451)
I think these features should be turned on by default, rather than off.

**Comment by Expertium** [comment]
> 2\. Don't print the init_w after pre-train. The user can wrongly assume that it is the optimized set of parameters.

I got confused by that too, it's definitely unnecessary.

**Comment by Expertium** [comment]
@L-M-Sherlock for some reason when I review new cards, the "Easy" interval is much higher than the corresponding S0 parameter. All decks have this problem.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/b744d0ae-a0cb-4f4f-bfc8-bc72c89e9779)

**Comment by Diplproveably** [comment]
> > This is needed to prevent the user from choosing insanely high or low values.
> 
> I agree. But the optimizer sometime suggests R=0.7. For some users with bad memory, they would not achieve 0.75
> 
> Besides, the scheduler would still store the input R. And the helper add-on should also cap it when getting the parameters from the custom scheduling code.

I think it's necessary to use a relatively low R in incremental writing; after all, it takes a relatively full dose of forgetting to allow for more creativity. So there's no need to put a limit on the R entered by the user.

**Comment by L-M-Sherlock** [comment]
> @L-M-Sherlock for some reason when I review new cards, the "Easy" interval is much higher than the corresponding S0 parameter. All decks have this problem.

What's your requested retention? And could you share the full configuration with me?

**Comment by L-M-Sherlock** [comment]
> I have some feedback to share about the optimizer. (mostly cosmetic changes)

Done in #350

**Comment by user1823** [comment]
@L-M-Sherlock
- Hide the graphs in section 2.1 of the optimizer by default but allow them to become visible if the user desires.
- The S0_rating_curve shouldn't execute at all if the initial stability for all the 4 ratings was calculated accurately (i.e. the data was adequate for all the 4 ratings). Executing it doesn't have any adverse effect. But, it makes no sense to run it even if we have already calculated the initial stability for all the ratings.
- There is a typo in the first code block:
> If you upload collection file, replace it with your ~~colpgk~~ colpkg filename.
- Just after the output parameters (in section 2.2), add a message: "Note: These values should be used with FSRS scheduler v4.0.0 or above." Also, make the "Note:" part bold and with red font color.

Reason for adding this message:
Let's say that a user didn't notice that a new major version has been released. Then, he used the optimizer for the monthly update of his parameters. Now, if he is not informed of a major change in the scheduler, he would put the parameters into the old scheduler (assuming that he didn't notice an increase in the number of parameters).
Then, he will come asking for help regarding the error shown by the scheduler (if it actually shows an error) or regarding incorrect scheduling of the cards (if no error is shown by the scheduler).

**Comment by user1823** [comment]
@L-M-Sherlock, just a small nit:
The note added with the last patch doesn't look right. Especially, the "Note" part looks odd (perhaps because of the colour). Also, the text is duplicated. It is there in the code as well as the output.

I recommend you to use the following in a text block just after the output parameters.
```html
<font color=orange>Note: These values should be used with FSRS scheduler v4.0.0 or above.</font>
```

Like this:
<img width=1000 src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/dd61b226-d990-4df7-9706-58420d4da610">

**Comment by user1823** [comment]
@L-M-Sherlock, one advice regarding the release process.

Release an update to the helper add-on (after it has been tested) before confirming the release for the scheduler and optimizer. This will ensure that most users would have the latest version of the helper add-on when they update the scheduler code.

**Comment by L-M-Sherlock** [comment]
> I recommend you to use the following in a text block just after the output parameters.

OK. I will pick up this suggestion in the next version. But I wouldn't release a patch version just for it.

> Release an update to the helper add-on (after it has been tested) before confirming the release for the scheduler and optimizer. This will ensure that most users would have the latest version of the helper add-on when they update the scheduler code.

Of course. I will release the scheduler and optimizer formally after we update the helper add-on and refine the documents.

**Comment by galantra** [comment]
> When I insert this new code "fsrs4anki_scheduler.js"
> 
> ![image](https://user-images.githubusercontent.com/42243532/253273902-46fd1adb-4462-4211-8c72-5c148dd61751.png)
> 
> and when running the helper this message appears below
> 
> ![image](https://user-images.githubusercontent.com/42243532/253273725-753235b4-e7e2-483d-8317-ca43eccd0c8a.png)

I used to get the same error (using Optimizer 4.0.1 and Scheduler qt6 v4.0.0) on 2 out of 4 collections that I tried.

Then I ran the optimizations again with the updated Optimizer v4.0.3, which yielded different parameters. These didn't result in any errors.

**Comment by L-M-Sherlock** [comment]
> I used to get the same error (using Optimizer 4.0.1 and Scheduler qt6 v4.0.0 and the Helper from the first post in this thread) on 2 out of 4 collections that I tried.

Could you provide the completed scheduler code affected by this error?

**Comment by Expertium** [comment]
> > @L-M-Sherlock for some reason when I review new cards, the "Easy" interval is much higher than the corresponding S0 parameter. All decks have this problem.
> 
> What's your requested retention? And could you share the full configuration with me?

Oh, I'm an idiot. My requested R is 0.8, of course the intervals are gonna be longer.

**Comment by galantra** [comment]
> > I used to get the same error (using Optimizer 4.0.1 and Scheduler qt6 v4.0.0 and the Helper from the first post in this thread) on 2 out of 4 collections that I tried.
> 
> Could you provide the completed scheduler code affected by this error?

I've just reverted to the configuration I used at the time and tried to reproduce the error. Thanks to https://github.com/sabrogden/Ditto/, I was even able to search in my clipboard history for the exact JS script I used.

Long story short, it was very likely just a syntax error. Because the JS script had two newlines after `"w":` that shouldn't have been there.

<details>
<summary>Relevant excerpt</summary>


// FSRS4Anki v4.0.0 Scheduler Qt6
set_version();
// The latest version will be released on https://github.com/open-spaced-repetition/fsrs4anki

// Configuration Start

const deckParams = [
  {
    // Default parameters of FSRS4Anki for global
    "deckName": "global config for FSRS4Anki",
    "w": 

[28.65, 29.51, 63.65, 1420.09, 4.5854, 0.1884, 1.1027, 0.1174, 1.686, 0.3616, 1.1866, 2.9257, 0.0174, 0.4557, 1.9994, 0.0001, 3.107],
    // The above parameters can be optimized via FSRS4Anki optimizer.
    // For details about the parameters, please see: https://github.com/open-spaced-repetition/fsrs4anki/wiki/Free-Spaced-Repetition-Scheduler
    // User's custom parameters for global
    "requestRetention": 0.9, // recommended setting: 0.8 ~ 0.9
    "maximumInterval": 36500,
    // FSRS only modifies the long-term scheduling. So (re)learning steps in deck options work as usual.
    // I recommend setting steps shorter than 1 day.
  },
  {
    // Example 1: User's custom parameters for this deck and its sub-decks.
    // Need to add <div id=deck deck_name="{{Deck}}"></div> to your card's front template's first line.
    "deckName": "ALL::Learning::English::Reading",
    "w": [1.2, 2.4, 3.6, 4.8, 5.1483, 1.4221, 1.2282, 0.035, 1.4668, 0.1286, 0.7539, 1.9671, 0.2307, 0.32, 0.9451, 0.5, 2],
    "requestRetention": 0.9,
    "maximumInterval": 36500,
  },
  {
    // Example 2: User's custom parameters for this deck and its sub-decks.
    // Don't omit any keys.
    "deckName": "ALL::Archive",
    "w": [1.3, 1.8, 2.3, 2.8, 4.9532, 1.502, 1.0922, 0.0081, 1.3771, 0.0294, 0.6718, 1.8335, 0.4066, 0.7291, 0.5517, 0.5, 2],
    "requestRetention": 0.9,
    "maximumInterval": 36500,
  }
];


</details>

When I removed the newlines, it worked.

**Comment by L-M-Sherlock** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/8299438e-caaf-447e-9552-2e9c2ae1fc42)

The helper add-on uses regex to match the values of weights, which can not match this case.

**Comment by nb9618** [comment]
I've got a minor suggestion. For me, uploading an exported deck to Colab often fails without any error description. What I have found after some searching, is that uploading a file using code in the notebook is for some reason much more reliable. I suggest adding a following message and a code block after line "You can export it via File -> Export... or Ctrl + E in the main window of Anki.":

> If you have problems uploading your file in Files panel, uncomment the following line and execute it:

    # from google.colab import files; _ = files.upload()

**Comment by user1823** [comment]
I have got a suggestion too.

Make the optimizer save the analysis for all the first ratings in the Files section of Google Colab (for e.g. how revlog.csv is saved).

It will help in debugging when any user complains that the initial stability doesn't look right. This is especially important because our approach for estimating the initial stability can result in wierd values for some types of data.

Adding the following line of code at the end of the code that does curve fitting for each grade worked for me.

```py
group.to_csv(f'Data for {first_rating}.csv', index=False)
```
However, it would be better to combine the files into a single CSV file instead of saving 4 different files.

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/cbd744de-6508-4ebd-b92c-fe6b286ee12f)
I don't think that these stats are necessary to display.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/30e1ca24-4390-4540-8eff-81f83157be18)
I think it would be better to have one big progress bar instead of 5 bars (one for each split).
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/e5c22d09-a1dd-4037-a638-7dc9d229e740)
These graphs should be removed as well. You can keep them in "dev" versions (versions where all of the code is visible and you can edit any of it), but they're just confusing for most users, I suggest only leaving the calibration graph.

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/a234a935-c634-4b34-a32a-6b3f649f633e)
Also, as I've mentioned, all 3 of these options should be on by defualt, rather than off. And remove "requires Fuzz", since many users don't even know about that line of code, so it's confusing.

**Comment by nb9618** [comment]
I think that the estimation for Easy initial stability for when there is not enough data is way too optimistic.
```
first rating: 1
rating history: 1,3,3,3,3,3,3,3,3,3,3
interval history: 0d,1d,3d,11d,1.2m,3.7m,10.5m,2.3y,5.9y,14.2y,32.5y
difficulty history: 0,5.7,5.5,5.4,5.2,5.1,5.1,5.0,5.0,4.9,4.9

first rating: 2
rating history: 2,3,3,3,3,3,3,3,3,3,3
interval history: 0d,1d,4d,14d,1.5m,4.7m,1.1y,2.9y,7.4y,17.7y,40.3y
difficulty history: 0,5.3,5.2,5.1,5.0,5.0,4.9,4.9,4.9,4.9,4.9

first rating: 3
rating history: 3,3,3,3,3,3,3,3,3,3,3
interval history: 0d,7d,25d,2.7m,8.1m,1.8y,4.8y,11.9y,27.8y,61.9y,100.0y
difficulty history: 0,4.8,4.8,4.8,4.8,4.8,4.8,4.8,4.8,4.8,4.8

first rating: 4
rating history: 4,3,3,3,3,3,3,3,3,3,3
interval history: 0d,10.2m,2.4y,6.3y,15.6y,36.4y,80.8y,100.0y,100.0y,100.0y,100.0y
difficulty history: 0,4.4,4.5,4.6,4.6,4.7,4.7,4.7,4.8,4.8,4.8
```
```
w = [0.7, 0.82, 6.85, 307.23, 4.8285, 0.4502, 0.503, 0.2325, 1.582, 0.1, 0.9801, 2.1102, 0.0901, 0.304, 1.0834, 0.4167, 2.2577]
```

There is no chance that I will remember something in ten months after a single review even if I answered "Easy". I think that the maximal value of w3 which I'd personally be comfortable with would be around 20.
The problem with this kind of overestimation is that the user may not notice that something is wrong and it will be revealed only after a very long period of time.

**Comment by Expertium** [comment]
In [the other issue](https://github.com/open-spaced-repetition/fsrs4anki/issues/282) I suggested a method that can decrease S0 for "Easy" and make it more realistic, and I also suggested making it so that it cannot exceed 365. We might limit it even more, to something like 30-40, but the problem is that it could hurt material that it genuinely _that_ easy.

**Comment by L-M-Sherlock** [comment]
Why not just limit the first interval in the scheduler?

**Comment by Expertium** [comment]
You still have to choose a reasonable max value either way.

**Comment by nb9618** [comment]
I don't think this will hurt very easy cards a lot, because in this case the user will review them and press easy a couple of times. This won't take much of their time, and the intervals would grow very quickly.

Here I set the first rating to Easy to simulate a lower initial stability.

```
rating history: 3,4,4,3,3,3,3,3,3
interval history: 0,7,50,312,901,2404,5986,14039,31233,36500
difficulty history: 0,4.8,4.4,4.1,4.3,4.4,4.5,4.6,4.6,4.7
```

**Comment by L-M-Sherlock** [comment]
I think 40 days is OK, because woz used 40 days as the max interval for https://supermemo.guru/wiki/Cloze_interval

**Comment by Expertium** [comment]
Ok then, let's add additive smoothing (it makes S0 for "Easy" a bit more realistic) to the final version and update the bounds for S0.
```
params, covs = curve_fit(power_curve, delta_t, recall, sigma=1/np.sqrt(count), bounds=((0.1), (40)))
rating_stability[rating] = max(min(S0_rating_curve(rating, *params), 40), 0.1)
```

**Comment by user1823** [comment]
> Why not just limit the first interval in the scheduler?

Limiting the interval without limiting the stability doesn't make sense. Reasons:
1. The next interval would still be very large (based on the next stability).
2. It disturbs the relationship between the stability, the requestRetention and the interval.

Also, as Expertium said, we would still need to think of an appropriate limit.

In my opinion, 60 days should be a reasonable upper limit for S0. Even if 60 days is an underestimate (and the actual value should have been about 200 days), it would still not increase the workload much.

A way to decrease the impact of this "forced" limit could be to remove the upper limit if the number of reviews with that first rating > 1000.

**Comment by Expertium** [comment]
> 1. The next interval would still be very large (based on the next stability).

That's a good catch, I haven't thought about that.

**Comment by szalejot** [comment]
@L-M-Sherlock 
Minor bug in the command line version of optimizer. It wasn't updated alongside with v4.0
In the `package/fsrs4anki_optimizer/__main__.py` file lines 85-86 are:
```
    optimizer.define_model()
    optimizer.train()
```
That kinda works, but the first 4 parameters are always `[1.0, 2.0, 3.0, 4.0`
I suppose it should rather be:
```
    optimizer.define_model()
    optimizer.pretrain(verbose=False)
    optimizer.train(verbose=False)
```
to produce correct results.

**Comment by L-M-Sherlock** [comment]
> I don't think that these stats are necessary to display.

> These graphs should be removed as well. You can keep them in "dev" versions (versions where all of the code is visible and you can edit any of it), but they're just confusing for most users, I suggest only displaying the calibration graph, the one without D and S.

They are useful debug info. I tend to keep them.

**Comment by L-M-Sherlock** [comment]
> I recommend you to use the following in a text block just after the output parameters.

> However, it would be better to combine the files into a single CSV file instead of saving 4 different files.

> EDIT: I like user1823's idea. If a user has >1000 reviews for some grade, then set the limit to 100 instead of 40. The rationale is that if the user has a lot of reviews, then we can lift the limit a little bit because the estimate of S will be very accurate.

> In my opinion, 60 days should be a reasonable upper limit for S0. Even if 60 days is an underestimate (and the actual value should have been about 200 days), it would still not increase the workload much.

> Minor bug in the command line version of optimizer. It wasn't updated alongside with v4.0

Done in #357

**Comment by Expertium** [comment]
> They are useful debug info. I tend to keep them.

As I said, feel free to keep them in "dev" versions where all code is visible, I just don't think they are useful for the average person in "official" versions. Just display one calibration graph.

**Comment by L-M-Sherlock** [comment]
> As I said, feel free to keep them in "dev" versions where all code is visible, I just don't think they are useful for the average person in "official" versions. Just display one calibration graph.

If our sole focus is on providing necessary information to users, then displaying parameters and suggested retention would suffice. But for me, I retaining the current information display aids in code maintenance. I wouldn't want users to have to revert to the dev version when reporting bugs.

**Comment by Expertium** [comment]
Ok, I see.
Some more things:
1) [This](https://github.com/open-spaced-repetition/fsrs4anki/issues/348#issuecomment-1636784231). Also, I suggested capping requested R in the scheduler/helper add-on. Previously I suggested 0.75-0.97, but if you think that for some users values <0.75 could be optimal, you can implement a different range.
2) user1823 suggests that we should implement additive smoothing for S0 in the final version, just without adding +1 in `count = group['y']['count']`, which I agree with. According to my testing additive smoothing improves RMSE by around 1%. The difference was statistically significant, but it could've been a false positive, so I'm not sure. Well, it definitely doesn't make things worse, so my opinion is "Why not?".

**Comment by L-M-Sherlock** [comment]
> 1. [This](https://github.com/open-spaced-repetition/fsrs4anki/issues/348#issuecomment-1636784231). Also, I suggested capping requested R in the scheduler/helper add-on. Previously I suggested 0.75-0.97, but if you think that for some users values <0.75 could be optimal, you can implement a different range.

In this case, should the helper add-on warn the user that the setting retention is out of range?

> 2. user1823 suggests that we should implement additive smoothing for S0 in the final version, just without adding +1 in `count = group['y']['count']`, which I agree with. According to my testing additive smoothing improves RMSE by around 1%. The difference was statistically significant, but it could've been a false positive, so I'm not sure. Well, it definitely doesn't make things worse, so my opinion is "Why not?".

OK. I will add it to v4 in the next patch.

**Comment by Expertium** [comment]
> In this case, should the helper add-on warn the user that the setting retention is out of range?

Yes. Here are 2 messages, for R>0.97 and for R<0.75 or whatever value you choose.

"Warning: requested retention should be between 0.75 and 0.97. Your requested retention is too high, it will lead to an enormous workload, so it has been set to 0.97."
"Warning: requested retention should be between 0.75 and 0.97. Your requested retention is too low, it will lead to a lot of material being forgotten, so it has been set to 0.75."

Also, I just got an idea for the helper-add on, but it's not directly related to to v4, so I'll make a new issue.

**Comment by galantra** [comment]
In 4.1.0, the notebook says that it can't find the module ‘fsrs4anki_optimizer’.

<details>
<summary>Error message (screenshot)</summary>

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/52420923/231fc9d1-9297-44de-b6e6-32f3a1fe06e0)
</details>

<details>
<summary>Error message (text)</summary>
```

---------------------------------------------------------------------------

ModuleNotFoundError                       Traceback (most recent call last)

[<ipython-input-6-05cb2ffe5175>](https://localhost:8080/#) in <cell line: 6>()
      4 import sys
      5 sys.path.insert(0, os.path.abspath('./package/fsrs4anki_optimizer'))
----> 6 import fsrs4anki_optimizer as optimizer
      7 optimizer = optimizer.Optimizer()
      8 optimizer.anki_extract(filename)

ModuleNotFoundError: No module named 'fsrs4anki_optimizer'


---------------------------------------------------------------------------
NOTE: If your import is failing due to a missing package, you can
manually install dependencies using either !pip or !apt.

To view examples of installing some common dependencies, click the
"Open Examples" button below.
---------------------------------------------------------------------------

```
</details>

**Comment by L-M-Sherlock** [comment]
My mistakes. I have fixed it just now.

**Comment by Expertium** [comment]
Sherlock, what's the reasoning behind this?
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/09f20c7f-9e42-476f-b863-8e23d901a9c8)
As I previously mentioned, in statistics the standard error of the mean is inversely proportional to the square root of the number of observations, not the number itself.

**Comment by moshe4512** [comment]
Hi, I only answer again or good, and never answer hard or easy. How will the optimizer work for that case? Also, when you get to the end of the optimizer and it is showing you the graphs for each answer choice, it stops at hard (cannot divide by zero- since there are no cases), and I don’t get to see the graph for good.

**Comment by L-M-Sherlock** [comment]
> Hi, I only answer again or good, and never answer hard or easy. How will the optimizer work for that case? Also, when you get to the end of the optimizer and it is showing you the graphs for each answer choice, it stops at hard (cannot divide by zero- since there are no cases), and I don’t get to see the graph for good.

Solved by #369.

**Comment by perhydrol** [comment]
> Your deck only has 400+ reviews, which are not enough to optimize the model. I will add some information about that instead of error raised by the code.

Thanks, I got it. So, should I continue to use v3 of fsrs4anki in this case or go back to Anki's default algorithm?

**Comment by Expertium** [comment]
You can continue using v3, though with 400 reviews, it won't be accurate either. Before releasing the new version, Sherlock will run v4 on many collections to find good average parameters, and then you can use v4 with those average parameters until you accumulate 1000+ reviews.

**Comment by user1823** [comment]
I don't know what https://github.com/open-spaced-repetition/fsrs4anki/pull/371 did to my collection. 

I got the following stability_for_pretrain.tsv with my collection that I last shared with you (about 2 weeks back). It doesn't have any data about first rating = 1. Also, several reviews for first rating = 3 have been filtered out. The reason that a large number of my cards have the first interval = 1d (with first rating = 3) is that 1d was my graduating interval when I was using the default Anki algorithm.

[stability_for_pretrain.tsv.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/12110244/stability_for_pretrain.tsv.zip) (Remove .zip from end)

For comparison, here is the stability for pretrain generated using v4.1.3 for the same collection:
[stability_for_pretrain.tsv.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/12110384/stability_for_pretrain.tsv.zip) (Remove .zip from end)

Also, with v4.2.0, I got crazy low RMSE (see below). This is one of the reasons I asked you to implement https://github.com/open-spaced-repetition/fsrs4anki/issues/337. Had this been implemented, RMSE would have indicated that the trained parameters would not work well for my "actual" collection.

```
R-squared: 0.9044
RMSE: 0.0056
MAE: 0.0044

Last rating: 1
R-squared: 0.6749
RMSE: 0.0231
MAE: 0.0118

Last rating: 2
R-squared: -3.6253
RMSE: 0.0936
MAE: 0.0875

Last rating: 3
R-squared: 0.8456
RMSE: 0.0070
MAE: 0.0050

Last rating: 4
R-squared: -84.9619
RMSE: 0.0458
MAE: 0.0416
```

**Comment by Expertium** [comment]
> I don't think outliers should be included in the calculation of evaluation metrics.

No, I agree with user1823 here. You can exclude outliers from training, but they are still there, and users will still encounter them in practice. We need to ensure that whatever we are doing makes the results better (on average) for the whole dataset rather than just for the part of the dataset that has no outliers. If some method makes results better when there sre no outliers, but makes results worse overall when outliers are present, then it's not a good method. Remember, you can't remove them during the actual execution of FSRS in Anki.

**Comment by Expertium** [comment]
Btw, Sherlock, I noticed that you included MAE in the stats.I thought you were against using MAE in favor of RMSE?

**Comment by L-M-Sherlock** [comment]
> No, I agree with user1823 here.

If we consider the outliers in evaluation, the graph shown in https://github.com/open-spaced-repetition/fsrs4anki/issues/348#issuecomment-1644059545 will cancel the more rational initial stability. But s=256.18 is ridiculous in that case. Could you explain it?

> Btw, Sherlock, I noticed that you included MAE in the stats. I thought you were against using MAE in favor of RMSE?

I am just curious about that metric.

**Comment by Expertium** [comment]
> Could you explain it?

Perhaps that person is studying outside of Anki more than "inside". In other words, he sees this material very often and as a result he already knows the material really well. I've tested a different sigma for S0 on [this](https://github.com/open-spaced-repetition/fsrs4anki/issues/282#issuecomment-1638580461) person's collection, and the more reasonable values actually increased RMSE, suggesting that what we think is "reasonable" doesn't work for that person.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/644a4191-0f73-466c-8c49-9978b0c05257)

**Comment by user1823** [comment]
> If we consider the outliers in evaluation, the graph shown in [#348 (comment)](https://github.com/open-spaced-repetition/fsrs4anki/issues/348#issuecomment-1644059545) will cancel the more rational initial stability. But s=256.18 is ridiculous in that case. Could you explain it?

I am not very sure what you mean by "cancel". But, even if the RMSE increases by including the outliers in evaluation, it makes sense because the model couldn't predict that those cards would be recalled after 400+ days just after one review. I agree that this is not a fault of the model, but the evaluation should still include all the reviews to be "fair".

**Comment by L-M-Sherlock** [comment]
> I am not very sure what you mean by "cancel".

I mean the worse RMSE would not support the rational prediction.

> I agree that this is not a fault of the model, but the evaluation should still include all the reviews to be "fair".

I think it depends. If those outliers are not representative, it is safe to exclude them. And it is meaningless to evaluate the performance in those outliers.

> Also, if the evaluation data is not the same for both the optimizations, the RMSEs are not comparable.

IMO, the outlier filter is independent from optimization. We can apply the filter before optimization.

**Comment by L-M-Sherlock** [comment]
> Perhaps that person is studying outside of Anki more than "inside". In other words, he sees this material very often and as a result he already knows the material really well. I've tested a different sigma for S0 on [this](https://github.com/open-spaced-repetition/fsrs4anki/issues/282#issuecomment-1638580461) person's collection, and the more reasonable values actually increased RMSE, suggesting that what we think is "reasonable" doesn't work for that person.

In his data, most outliers have lengthy intervals, likely due to backlog. I guess the story is: Usually, users are inclined to assign high rating while handling backlog, as it aids in clearing the backlog faster. This inclination leads to inflated retention rates for reviews with long intervals. So filtering out these outliers (fake reviews) is rational. And it is unnecessary to evaluate the model with these outliers.

**Comment by L-M-Sherlock** [comment]
In #374, I calculate 30+ set of parameters from datasets collected in https://forms.gle/KaojsBbhMCytaA7h8. I use the median for the new default global parameters. It's time to release the stable version for FSRS v4.

**Comment by user1823** [comment]
It is somewhat concerning that the choice of the initial set of parameters (https://github.com/open-spaced-repetition/fsrs4anki/commit/6559dc359cfe46653d0fec1b00643b00a95e1684) has a noticeable impact (4.4%) on the RMSE even for @L-M-Sherlock's collection, which has a significant number of reviews.

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/fc81a7dd-dfd2-45a7-8fca-5dcaacd1ecd7)

**Comment by user1823** [comment]
@L-M-Sherlock, don't you think that 15 days as the maximum limit of initial stability (when total_count < 1000) is too small? For comparison, the initial stability for Good in my collection is 15.84 days. I think that the maximum limit should be at least 40 days, though 60 days is also fine.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/91272ed1-b972-47d1-96d4-f8902e087c64)
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/9ec1db1d-b3c8-4e02-aa75-6788416f2766)

**Comment by Expertium** [comment]
I haven't seen that change. 
@L-M-Sherlock I agree with user1823, 15 days is way too small. I suggest 40.

**Comment by L-M-Sherlock** [comment]
I'm afraid that 60 days would scare new users.

**Comment by Expertium** [comment]
How about 30?

**Comment by L-M-Sherlock** [comment]
OK. I will update it in the next patch.

**Comment by L-M-Sherlock** [comment]
> It is somewhat concerning that the choice of the initial set of parameters ([6559dc3](https://github.com/open-spaced-repetition/fsrs4anki/commit/6559dc359cfe46653d0fec1b00643b00a95e1684)) has a noticeable impact (4.4%) on the RMSE even for @L-M-Sherlock's collection, which has a significant number of reviews.

Maybe it is due the sensitivity of RMSE? As I mentioned in https://github.com/open-spaced-repetition/fsrs4anki/issues/342#issuecomment-1630367207, the RMSE for all last ratings decreased. But the total RMSE increased.

**Comment by user1823** [comment]
> Maybe it is due the sensitivity of RMSE? As I mentioned in [#342 (comment)](https://github.com/open-spaced-repetition/fsrs4anki/issues/342#issuecomment-1630367207), the RMSE for all last ratings decreased. But the total RMSE increased.

But, my point is that the final parameters (and thus, the RMSE) should be the same regardless of what initial parameters are used if the data is sufficient.

**Comment by user1823** [comment]
Also, the latest optimizer filters out my data for first rating = 3 and delta_t = 4 (and above). But, the initial stability for Good for my collection is more than 10 days (exact value depends on which version of the optimizer I use). This means that the optimizer is not using my newer data for training (unless the count with high delta_t eventually becomes so large that it fits into the threshold). 

<img width="501" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/0d96b23d-6e66-46e4-a70a-212b85b096f4">

As I said before, the reason that a large number of my cards (3944) have the first interval = 1d (with first rating = 3) is that 1d was my second learning step when I was using the default Anki algorithm. But, this should not be a factor to decide how well the optimizer works for me.

**Comment by Expertium** [comment]
Do you mean that there might be multiple sets of parameters that achieve the minimum possible values of the loss function? If so, then it really doesn't seem that way to me. In neural networks, yes, multiple equivalent minima are possible. But in neural networks, all neurons are identical, it's not like each neuron has its own special feature. In FSRS, each formula is different, and one parameter cannot replace the other. For example, if you switched w[11] and w[14] in the code below:
```
        new_s = self.w[11] * \
                torch.pow(new_d, -self.w[12]) * \
                (torch.pow(state[:,0] + 1, self.w[13]) - 1) * \
                torch.exp((1 - r) * self.w[14])
```
the result would be nonsensical. So from a theoretical point of view, it seems that in FSRS there can only be one global minima, in other words, there can only be one specific set of parameters that achieves the minimum possible value of the loss function, there cannot be multiple "equivalent" sets.

**Comment by L-M-Sherlock** [comment]
> Do you mean that there might be multiple sets of parameters that achieve the minimum possible values of the loss function? If so, then it really doesn't seem that way to me. In neural networks, yes, multiple equivalent minima are possible. But in neural networks, all neurons are identical, it's not like each neuron has its own special feature. In FSRS, each formula is different, and one parameter cannot replace the other.

But FSRS is not a linear model. There are many non-linear components in FSRS. So it is possible that the optimization has several minimum.

**Comment by Expertium** [comment]
> We can increase the batch size epoch by epoch. When the batch size is equal to the size of training set, the gradient of loss function will point to the global minimum.

That's interesting, I would like to see it and test.

**Comment by Expertium** [comment]
So Sherlock, what are you working on right now?

**Comment by L-M-Sherlock** [comment]
I'm on a business trip, so not available in this week.

**Comment by L-M-Sherlock** [comment]
By the way, I asked my schoolmate major in Computational Mathematics whether the optimizer could reach the global minimum in FSRS. He mentioned that the optimization for multivariable function (in FSRS) is very hard to analysis. For example, the [saddle point](https://byjus.com/maths/saddle-points/) is the point where the gradient is zero. But it is not a local/global minimum. Not any optimization algorithm could ensure that the optimizer could find out the global minimum in FSRS.

**Comment by Expertium** [comment]
@L-M-Sherlock I think in the current version of the optimizer it's possible that a value for "Good" will be larger than for "Easy" if "Good" has more datapoints.
`params, _ = curve_fit(power_forgetting_curve, delta_t, recall, sigma=1/np.sqrt(count), bounds=((0.1), (30 if total_count < 1000 else 365)))`
You should probably add some kind of extra cap to ensure that S0 for "Good" cannot be greater than S0 for "Easy" even if `total_count` is greater than 1000 for "Good" and less than 1000 for "Easy".
